File: Keith Ho/week_4/day_2/mta_lab.rb

@subway = [
  {
    "name" => "N",
    "stops" => ["Times Square", "34th", "28th", "23rd", "Union Square", "8th"]
  },
  {
    "name" => "L",
    "stops" => ["8th", "6th", "Union Square", "3rd", "1st"]
  },
  {
    "name" => "6",
    "stops" => ["Grand Central", "33rd", "28th", "23rd", "Union Square", "Astor Place"]
  }
]

def arr_of_lines line
  i = 0
  while i < @subway.size
    if @subway[i]["name"] == line
      return @subway[i]["stops"]
    end
    i += 1
  end
end

def index_of_stop line, stop
  i = 0
  @a = 0
  while i < @subway.size
    if @subway[i]["name"] == line
      while @a < @subway[i]["stops"].size
        if @subway[i]["stops"][@a] == stop
          return @a
        end
        @a += 1
      end
    end
    i += 1
  end
end

def stops_between line, start, final
  arr_line = arr_of_lines(line)
  index_of_start = index_of_stop(line, start).to_i
  index_of_final = index_of_stop(line, final).to_i

  arr_stops_in_order = []
  @arr_stops = []

  if index_of_start > index_of_final
    if index_of_start - index_of_final > 1
      @arr_stops = arr_line[index_of_final + 1, (index_of_start - index_of_final - 1)]
      i = @arr_stops.size
      while i >= 0
        arr_stops_in_order.push @arr_stops[i]
        i -= 1
      end
      @arr_stops = arr_stops_in_order[1, arr_stops_in_order.size]
    else
      @arr_stops = []
    end
  else
    if index_of_final - index_of_start > 1
      @arr_stops = arr_line[index_of_start + 1, (index_of_final - index_of_start - 1)]
    else
      @arr_stops = []
    end
  end

  return @arr_stops
end

def plan_trip start_line, start_stop, end_line, end_stop
  stops_to_union = stops_between(start_line, start_stop, "Union Square")
  stops_from_union = stops_between(end_line, "Union Square", end_stop)

  if start_stop == "Union Square"
    total_stops = stops_to_union.size + stops_from_union.size + 1
  else
    total_stops = stops_to_union.size + stops_from_union.size + 2
  end

  @message_different_lines = "You must travel through the following stops on the #{ start_line } line:
          #{ stops_to_union.join(", ") } Union Square.
          Change at Union Square.
          Your journey continues through the following stops:
          #{ stops_from_union.join(", ") } #{ end_stop }.
          #{ total_stops } stops in total."

  @message_same_line = "You must travel through the following stops on the #{ start_line } line:
          #{ stops_to_union.join(", ") }, Union Square, #{ stops_from_union.join(", ") }, #{ end_stop }.
          #{ total_stops } stops in total."

  @message_union_square = "You must travel through the following stops on the #{ end_line } line:
          #{ stops_from_union.join(", ") }, #{ end_stop }.
          #{ total_stops } stops in total."

  if start_line != end_line
    if start_stop == "Union Square"
      puts @message_union_square
    else
      puts @message_different_lines
    end
  elsif
    if start_stop == "Union Square"
      puts @message_union_square
    else
      puts @message_same_line
    end
  end
end

def menu
  puts "Trip Planner".center(50)
  puts "*" * 40
  puts "[s] - Start"
  puts "[q] - Quit"
  puts "Enter your choice:"
end

menu
menu_choice = gets.chomp.downcase

until menu_choice == "q"
  case menu_choice
  when "s"
    puts "L: 1st, 3rd, Union Square, 6th, 8th"
    puts "N: Times Square, 34th, 28th, 23rd, Union Square, 8th"
    puts "6: Astor Place, Union Square, 23rd, 28th, 33rd, Grand Central"
    puts "-" * 20
    puts "Enter your start line:"
    startline = gets.chomp.capitalize
    puts "Enter your start stop:"
    startstop = gets.chomp.capitalize
    puts "Enter your end line:"
    endline = gets.chomp.capitalize
    puts "Enter your end stop:"
    endstop = gets.chomp.capitalize

    plan_trip(startline, startstop, endline, endstop)

  else
    puts "Invalid selection. Choose again."
  end


  menu
  menu_choice = gets.chomp.downcase
end

puts "Thanks for using this Trip Planner!"

File: Kundai Khuleya/week_1/MTALAB/js/mtalab.js
//
// # MTA Lab
//
// ## Objectives:
// * Apply your knowledge of Javascript to solve a real world problem.
// * Get really good at array manipulation.
//
// #### Activity
// * Create a program that models a simple subway system.
//
// * The program takes the line and stop that a user is getting on at and the line
// and stop that user is getting off at and prints the journey and the total number of stops for the trip in the console:
//
// ```javascript
// planTrip('N', 'Times Square', '6', '33rd'); // This is only a suggested function name and signature.
//
// // console.log() shows output similar to this:
// // "You must travel through the following stops on the N line: 34th, 28th, 23rd, Union Square."
// // "Change at Union Square."
// // "Your journey continues through the following stops: 23rd, 28th, 33rd."
// // "7 stops in total."
// ```
//
// * There are 3 subway lines:
//   * The N line has the following stops: Times Square, 34th, 28th, 23rd, Union Square, and 8th
//   * The L line has the following stops: 8th, 6th, Union Square, 3rd, and 1st
//   * The 6 line has the following stops: Grand Central, 33rd, 28th, 23rd, Union Square, and Astor Place.
//   * All 3 subway lines intersect at Union Square, but there are no other intersection points. (For example, this means the 28th stop on the N line is different than the 28th street stop on the 6 line, so you'll have to differentiate this when you name your stops in the arrays.)
// * Tell the user the number of stops AND the stops IN ORDER that they will pass through or change at.
//
//
// #### Hints:
// * Work out how you would do it on paper first! Then start to explain that process in Javascript.
// * Get the program to work for a single line before trying to tackle multiple lines.
// * Don't worry about prompting the user for input. Hard code some values to get it working. You can use ```prompt()``` later to make it more interactive.
// * Consider diagramming the lines by sketching out the subway lines and their stops and intersection.
// * The key to the lab is finding the index positions of each stop. (hint: ```indexOf()```)
// * Make sure the stops that are the same for different lines have different names (i.e. 23rd on the N and on the 6 need to be differentiated)

console.log(createRoute("Change at Union Square."));

const lineData = {
    "N": [
        "Times Square",
        "34th",
        "28th",
        "23rd",
        "Union Square",
        "8th"
    ],
    "L": [
        "8th",
        "6th",
        "Union Square",
        "3rd",
        "1st"
    ],
    "6":[
        "Grand Central",
        "33rd",
        "28th",
        "23rd",
        "Union Square",
        "Astor Place"
    ]
};

const createRoute = function(start, end){
    let result = []
    // Determine if a change will be required (are the start and end on the same line)
    if(start.line === end.line){
        const line = lineData[start.line] // Extract line information
        const start_index = line.findIndexOf(start.station) // Get the index of the starting point
        const end_index = line.findIndexOf(end.station) // Get the index of the end point

        if(start_index > 0 && end_index > 0){ // Check that they are valid station
            if(start_index - end_index < 0){ // If the starting point comes first
                result = line.slice(start_index, end_index + 1)
            } else { // If not, then reverse
                result = line.reverse().slice(line.length - end_index, line.length - start_index)};
                result = [result.map(x => { // Format the output so that it can be parsed by report generator
                              return {
                                  "line": start.line,
                                  "station": x
                              }
                          })]
                      }
                  } else {
                      let result = []
                      const start_line = line_data[start.line]
                      const end_line = line_data[end.line]

                      let start_route = []
                      let end_route = []

                      // First, find the start station and route to Union Square
                      start_index = start_line.findIndexOf(start.station)
                      start_union_index = start_line.findIndexOf("Union Square")

                      if(start_index > 0){ // Check that the station is valid
                          if(start_index - start_union_index < 0){
                              start_route = start_line.slice(start_index, start_union_index + 1)
                          } else {
                              start_route = start_line.reverse().slice(start_line.length - start_union_index, start_line.length - start_index) // [TODO] Tweak this line as well.
                          }
                      } else { // Throw error if the station doesn't exist
                          throw Error("Invalid starting station.")
                      };
                      subarray

                            start_route = start_route.map(x => {
                                return {
                                    "line": start.line,
                                    "station": x
                                }
                            })

                            end_route = end_route.map(x => {
                                return {
                                    "line": end.line,
                                    "station": x
                                }
                            })

                            result = [start_route, end_route]
                        }

                        return result;
                    }

                    const main = () => {


                        const route = create_route(start, end)
                   if(route.length < 2){
                            route_stops = route[0].map(x => {x.station}).join(",")
                            console.log(`You must travel through the following stop on the ${route[0][0].line} line: ${route_stops}`)
                        } else {
                            console.log(`You must travel through the following stops on the ${route[0][0].line} line: ${route[0].map(x => x.station).join(",")}`)
                            console.log(`Change at Union Square`)
                            console.log(`Your journey continues through the following stops: ${route[1].map(x => x.station).join(",")}`)
                        }
                    }

                    main();

                    //¯\_(ツ)_/¯//   
